src/sleep/sleep.c

```diff
@@ -135,6 +137,82 @@ static int execute(char **modes, char **states) {
         return r;
 }
 
+static int read_wakealarm(uint64_t *result) {
+        _cleanup_free_ char *t = NULL;
+
+        if (read_one_line_file("/sys/class/rtc/rtc0/since_epoch", &t) >= 0)
+                return safe_atou64(t, result);
+        return -EBADF;
+}
+
+static int write_wakealarm(const char *str) {
+
+        _cleanup_fclose_ FILE *f = NULL;
+        int r;
+
+        f = fopen("/sys/class/rtc/rtc0/wakealarm", "we");
+        if (!f)
+                return log_error_errno(errno, "Failed to open /sys/class/rtc/rtc0/wakealarm: %m");
+
+        r = write_string_stream(f, str, 0);
+        if (r < 0) {
+                log_debug_errno(r, "Failed to write '%s' to /sys/class/rtc/rtc0/wakealarm: %m",
+                str);
+                return r;
+        }
+
+        return r;
+}
+
+static int execute_s2h(char **hibernate_delays, char **suspend_states) {
+
+        _cleanup_strv_free_ char **hibernate_modes = NULL, **hibernate_states = NULL;
+        char **delay_str;
+        uint64_t orig_time, cmp_time, delay;
+        char time_str[DECIMAL_STR_MAX(orig_time)];
+        int r;
+
+        r = parse_sleep_config("hibernate", &hibernate_modes, &hibernate_states);
+        if (r < 0)
+                return r;
+
+        STRV_FOREACH(delay_str, hibernate_delays) {
+                r = safe_atou64(*delay_str, &delay);
+                if (r < 0)
+                        return log_error_errno(errno, "Failed to read hibernate delay %s", *delay_str);
+        }
+
+        r = read_wakealarm(&orig_time);
+        if (r < 0)
+                return log_error_errno(errno, "Failed to read time: %d", r);
+
+        orig_time += delay * 60;
+        sprintf(time_str, "%" PRIu64, orig_time);
+
+        r = write_wakealarm(time_str);
+        if (r < 0)
+                return r;
+
+        r = execute(NULL, suspend_states);
+        if (r < 0)
+                return r;
+
+        r = read_wakealarm(&cmp_time);
+        if (r < 0)
+                return log_error_errno(errno, "Failed to read time: %d", r);
```

Comment:
Does this generate any entries in logs by default? I think it'd be nice to have a log line for each of the steps here.

Reply:
I was seeing entries when I looked at journalctl for the suspend-to-hibernate unit (basically when suspend is invoked, when it's woken and when hibernate is invoked and woken). Were you expecting them somewhere else or something more verbose?

Reply:
I was really an honest question. If there are logs, that's good.

Reply:
Thanks.
In my checking the one area that I think was needed still was I've added some extra `log_debug()` events for the sleep and wakeup times.